daemon/inertia/auth/password.go

```diff
@@ -0,0 +1,59 @@
+package auth
+
+import (
+	"errors"
+
+	"golang.org/x/crypto/bcrypt"
+)
+
+var (
+	errSameUsernamePassword = errors.New("Username and password must be different")
+	errInvalidUsername      = errors.New("Only letters, numbers and underscore are allowed in usernames")
+	errInvalidPassword      = errors.New("Only letters, numbers and underscore are allowed in passwords, and password must be at least 5 characters")
+)
+
+func hashPassword(password string) (string, error) {
+	hash, err := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
+	if err != nil {
+		return "", errors.New("bcrypt password hashing unsuccessful: " + err.Error())
+	}
+	return string(hash), nil
+}
+
+func correctPassword(hash string, password string) bool {
+	return bcrypt.CompareHashAndPassword([]byte(hash), []byte(password)) == nil
+}
+
+func validateCredentialValues(username, password string) error {
+	if username == password {
+		return errSameUsernamePassword
+	}
+	if len(password) < 5 {
```

Comment:
also validate username such that it cannot be empty?

Reply:
should we also impose a maximum on password and username to prevent overflows?

Reply:
Good point on making sure the username isn't empty! I'll fix that

As for imposing a maximum on credentials, I feel like if there really is an overflow, it would already be a problem before we reach this point, when we try to parse the request body - that said, perhaps there should be a maximum regardless. 100 characters? 🤔

Reply:
Yeah in all likelihood it would crash something else, but I think it's good design to not rely on that. Possibly set it to 128 as per [owasp]?(https://www.owasp.org/index.php/Authentication_Cheat_Sheet#Password_Length)

Also related discussion on passwords: [stackoverflow](https://stackoverflow.com/questions/98768/should-i-impose-a-maximum-length-on-passwords)

Reply:
I suppose it might depend on the hashing algorithm - I'll look into it, but 128 seems like a reasonable limit. Thanks for pointing this out!